feat: add support for default handlers

NeonOxide opened the pull request:
Added the wrapping of default handlers.

This PR contains:

- Removal of WrapInTests as that is now covered by this more general implementation
- Added wrapping around entry points of all necessary default handlers
- Modified the shell to handle non primitive effects by replacing a mock unit value with a call to a local definition. This is made to make :eval work.
- Modified Assert.runWithIO to be a default handler

Important questions yet to be answered before merging:

- Is the way we address the shell issue the way we want to do it?
- Right now we call eval on the entry point signature and decide which entry points to add based on that. If there is an error on the signature we just add all of them, is this the behavior we want?
- Right now the syntax coloring of the entry point effects does not work. It is not a great issue, it is only that the effects appear as white instead of with effect colors. If I had to guess it is because we are removing them. It is something important to fix before merging, but it is not all mangled. I am not familiar with the internals of the vscode code so I do not know what to pass along, right now I am trying to keep the original effect location as synthetic but I don't know what else to do. 

Example code if someone wants to try:
```flix
pub eff E{
   def op(): Unit
}
mod E {
    @DefaultHandler
    pub def runWithIO(f: Unit -> a \ ef): a \ (ef - E) + IO =
            run {
                f()
            } with handler E {
                def op(k) = {
                    println("Default behaviour");
                    k()
                }
            }
}
@Test
def myTest1(): Unit \ Assert = Assert.assertTrue(true)

@Test
def myTest2(): Unit \ Assert = Assert.assertTrue(false)

def main(): Unit \ E = E.op()
```
related to #11637

NeonOxide commented:
!!!jar

magnus-madsen commented on main/src/ca/uwaterloo/flix/language/phase/EntryPoints.scala:
```diff
@@ -96,6 +93,10 @@ object EntryPoints {
             (root.copy(mainEntryPoint = Some(entryPoint.sym)), Nil)
         }
       case Some(sym) => root.defs.get(sym) match {
+        case Some(shell) if shell.sym.name == "shell1" =>
```
Where is the string "shell1" used? I think we should put it into a constant.

NeonOxide replied:
The problem with shell1 is that it is also used as a magical string in Shell.scala. So if we want to unify their use so that changing one would change the other without breaking code we need to decide where to put this constant.

I think, a public static const in Shell.scala could be a good place to put it. Something like shellMainName or similar.

magnus-madsen commented on main/src/ca/uwaterloo/flix/language/phase/EntryPoints.scala:
```diff
@@ -106,6 +107,43 @@ object EntryPoints {
   }
 
+  /**
+    * Returns a new shell function that swaps the unit value for a call to _f.
```
I am confused by this statement. What unit value?

NeonOxide replied:
Addressed below too, I will resolve this one and keep the conversation in the other one

magnus-madsen commented on main/src/ca/uwaterloo/flix/language/phase/EntryPoints.scala:
```diff
@@ -106,6 +107,43 @@ object EntryPoints {
+    *   def main(): Unit \ expEffs + IO =
+    *     def _f()={ println(exp) }
+    *     _f()
+    * }}}
```
I think I understand better now. Maybe you want to start the documentation by saying we have this code snippet and we want to transform it to this other snippet.

magnus-madsen commented on main/test/ca/uwaterloo/flix/language/phase/TestEntryPoints.scala:
```diff
@@ -511,4 +511,28 @@ class TestEntryPoints extends AnyFunSuite with TestUtils {
+        |def main(): Unit \ E = E.op()
+        |""".stripMargin
+    val result = compile(input, Options.TestWithLibMin)
+    expectSuccess(result)
```
We also want a lot more tests:

- One default handler
- Two default handlers
- Three default handlers

magnus-madsen replied:
Checking whether it compiles correctly or not is already a great test. We have lots of those.

For checking "the output", you can just use assertEq.

As for having the IO+effect. Your handler can be pure, and just do if (false) print etc.

magnus-madsen commented on main/src/ca/uwaterloo/flix/runtime/shell/Shell.scala:
```diff
@@ -312,9 +312,8 @@ class Shell(bootstrap: Bootstrap, options: Options) {
         // Cast to allow any subset of the primitive effects
```
We should probably update the documentation here, and insert a reference to EntryPoints.

magnus-madsen commented on main/src/ca/uwaterloo/flix/language/phase/EntryPoints.scala:
```diff
@@ -639,6 +674,99 @@ object EntryPoints {
     }
+
+    /**
+      * Wraps an entry point function with all necessary default effect handlers.
```
Thinking about this, perhaps a concrete example is better than an abstract example.

After all, the code has the abstract case.

NeonOxide replied:
In this case it means it only wraps the effects that appear in the signature. For example with default handlers for A, B and C. If main only has A and B, then it will only wrap the inner expression in the default handlers for A and B. But I agree that a more concrete example would help clarify more. I will rewrite it and add a concrete example.

magnus-madsen commented on main/src/ca/uwaterloo/flix/language/phase/EntryPoints.scala:
```diff
@@ -639,6 +674,99 @@ object EntryPoints {
+        case Result.Err(_) => CofiniteSet.universe
+      }
+      val necessaryHandlers = defaultHandlers.filter(h => defEffects.contains(h.handledSym))
+      necessaryHandlers.foldLeft(currentDef)((defn, handler) => wrapInHandler(defn, handler, root))
```
We should say something about the order. Even if what we say is just that: We do not use any specific order.

magnus-madsen commented on main/src/ca/uwaterloo/flix/language/phase/EntryPoints.scala:
```diff
@@ -639,6 +674,99 @@ object EntryPoints {
+      * Wraps an entry point function with a default effect handler.
+      *
+      * Transforms a function:
+      * def f(arg1: tpe1, ...): tpe \ ef = exp
```
code blocks and indent

magnus-madsen commented on main/src/ca/uwaterloo/flix/language/phase/EntryPoints.scala:
```diff
@@ -639,6 +674,99 @@ object EntryPoints {
+      * - Updates the function's type signature accordingly
+      *
+      * @param defn        The entry point function definition to wrap
+      * @param handlerInfo Information about the default handler to apply
```
Please nothing called "info"

NeonOxide replied:
I did not notice. This code was partially done from before and it kept the name of the previous type. I will change it to be in line with the new type name.

magnus-madsen reviewed (commented):
I think this is pretty good 👍 

Overall, we want more tests and we want some polish (as usual).

magnus-madsen commented:
!!!jar

magnus-madsen commented:
@NeonOxide Would it be possible to provide a snippet completion for default handlers? Take a look at e.g. ExprSnippetCompleter

NeonOxide commented on main/src/ca/uwaterloo/flix/language/phase/EntryPoints.scala:
```diff
@@ -106,6 +107,43 @@ object EntryPoints {
+    *
+    * The set of effects with default handlers is not known from the start unlike the set of primitive effects.
+    * In order to support the current way of creating a shell (create some simple wrapping code and execute it),
+    * we need to substitute PrimitiveEffs with the real set of effects generated by exp at EntryPoints.
```
Yeah. However, I think we maybe would need NonDet soon too. I noticed Random requires NonDet and Random seems like one of the prime users of default handlers I would say.

NeonOxide commented on main/src/library/Assert.flix:
```diff
@@ -182,6 +182,7 @@ mod Assert {
     ///
     /// In other words, re-interprets the `Assert` effect using the `IO` effect.
     ///
+    @DefaultHandler
```
I think there are many other ones, mainly the ones that are related to interacting with the OS or terminal. So Logger, Console, FileRead, FileWrite, the network ones, etc. I already mentioned on another comment that I think Random should maybe be included, but I am also not sure (although right now with only IO it is not possible). 

However, there are so many effects that I think should be marked that I think it is better to do on another PR once this is merged. I did Assert because if not the tests would have stopped working.  What do you think?

NeonOxide commented on main/src/ca/uwaterloo/flix/language/phase/EntryPoints.scala:
```diff
@@ -639,6 +674,99 @@ object EntryPoints {
+      * Wraps an entry point function with all necessary default effect handlers.
+      *
+      * Transforms a function:
+      * def f(arg1: tpe1, ...): tpe \ ef = exp
```
Yes, will also do

NeonOxide commented on main/src/ca/uwaterloo/flix/language/phase/EntryPoints.scala:
```diff
@@ -639,6 +674,99 @@ object EntryPoints {
+      */
+    private def wrapInHandler(defn: TypedAst.Def, handlerInfo: DefaultHandler, root: TypedAst.Root)(implicit flix: Flix): TypedAst.Def = {
+      // Create synthetic locations
+      val syntheticEffLocation = defn.spec.eff.loc.asSynthetic
```
Okay, I will change them. However, I do not know how this will end up looking after (hopefully) fixing the VsCode coloring issues.

NeonOxide commented:
> There is some problem with source locations. Notice that the types here are not colored:
> <img alt="image" width="536" height="55" src="https://private-user-images.githubusercontent.com/2287953/511672813-36634e27-718a-4f4f-9155-0aaa56b5315b.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjI5MDQyMzksIm5iZiI6MTc2MjkwMzkzOSwicGF0aCI6Ii8yMjg3OTUzLzUxMTY3MjgxMy0zNjYzNGUyNy03MThhLTRmNGYtOTE1NS0wYWFhNTZiNTMxNWIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MTExMSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTExMTFUMjMzMjE5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NmEwN2U5MWZhMTE5YTM4NTVmYzY1YzhhM2E5Nzc1OGU4YTkxYzNjYmFjNWUxZmRkOTk2NzUwMzEyOWQ2ZDhlMCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.LnrnupynsAySzzKak2FszqSuEt0gbPFbfeMlaGPJ2A8">

It is weird. I also get coloring issues as I stated at the end of the original PR, but I get different coloring issues. I will try and find the root issue with it or at least a way to solve it.

NeonOxide commented:
Apparently some server at GitHub does not like me pushing as I am receiving Internal Server error 500 when trying to push my new changes. I will leave the changelog here and hopefully later I will be able to upload them after I am free again:


- Fixed documentation issues
- Added new tests
- Added some comments to the comments present in the PR
- Extracted the shell name to a constant
- Noticed that the current mkShell does not give the best errors as it just replaces right now checked_ecast(()) with println(s) which makes the locations a little bit wonky. When there is an error, it only prints the println(s) and then says that there is an effect error. Which there is, but it should show something more, I think I can solve this I just need to think about it a little bit.


I tried to fix the source location issues and found the following issues:
 + I am unable to reproduce your @magnus-madsen  specific coloring issues. The handlers for me are colored perfectly except for the first argument of the arrow type not being colored. But I get that on the current live version of the compiler too. 
 + The coloring issues I get are for the effects in the signature of the entry points, and I have no idea how to solve them nicely. The main issue is that we are removing effects from the signature, which then the lsp does not like because it does not match what the file has. So, if we just use SourceLocation.Unknown we get coloring back, but we cannot control click to jump to the effects and we get horrible error messages. And if we don't, we get "good" error messages but no coloring nor jump to. So far I think the only way of fixing this is that instead of modifying the function itself, we just create a new one and remove the old one from the entry points, but I do not know how feasible this is.

magnus-madsen commented:
I suggest we merge this and then a follow-up PR can move the transformation to `Lowering`

NeonOxide commented:
I also added some tests in Test.DefaultHandler.flix I tried to follow the guidelines as best as I could but maybe I missed something, so that could be something worth looking at.

magnus-madsen commented:
@mlutze Could you take a quick look also? 

After merging this, we will refactor and move the wrapping to Lowering (or something like that). 

But since that involves more changes, we were planning to merge this first.

magnus-madsen commented on main/test/flix/Test.DefaultHandler.flix:
```diff
@@ -0,0 +1,485 @@
+mod DefaultHandler {
```
mod Test.DefaultHandler

NeonOxide commented on main/src/ca/uwaterloo/flix/language/phase/EntryPoints.scala:
```diff
@@ -96,6 +94,10 @@ object EntryPoints {
       case Some(sym) => root.defs.get(sym) match {
+        case Some(shell) if shell.sym.name == Shell.ShellEntryPointName =>
+          // A main is given and it is shell - transform it.
+          val newShell = mkShell(shell)
```
Much more descriptive I agree

mlutze commented on main/src/ca/uwaterloo/flix/language/phase/EntryPoints.scala:
```diff
@@ -545,8 +584,11 @@ object EntryPoints {
-      (root, errors.toList.flatMap(_.toList) ++ duplicatedHandlerErrors)
+      val defsWithDefaultHandlers = ParOps.parMapValues(root.defs)(
+        defn =>
+          if (isEntryPoint(defn)(root)) wrapDefWithDefaultHandlers(defn, validHandlers, root) else defn
```
It looks like this transforms all entry points, but suppose I call a function somewhere in my code which happens to be an entry point. It should really only be transformed where it's _used_ as an entry point, right?

magnus-madsen commented on main/src/ca/uwaterloo/flix/language/phase/EntryPoints.scala:
```diff
@@ -640,6 +682,111 @@ object EntryPoints {
+      val defEffects: CofiniteSet[Symbol.EffSym] = eval(currentDef.spec.eff) match {
+        case Result.Ok(s) => s
+        // This means eff is either not well-formed or it has type variables.
+        // Either way, in this case we will wrap with all default handlers to provide better error messages
```
But the wrapper code will be moved to e.g. Lowering-- and those phases cannot report errors.

NeonOxide commented on main/test/ca/uwaterloo/flix/language/phase/TestEntryPoints.scala:
```diff
@@ -511,4 +511,268 @@ class TestEntryPoints extends AnyFunSuite with TestUtils {
+        |def main(): Unit \ E1 + E2 = E1.op();E2.op()
+        |""".stripMargin
+    val result = compile(input, Options.TestWithLibMin)
+    expectSuccess(result)
```
To be more clear. Technically right now they do not add anything extra. Right now all entry points are wrapped in the same way without any distinction. This is more of a future proof test to ensure that this invariant holds after changes.

magnus-madsen commented:
Merged! Thanks-- next up: Move to Lowering
